### PR TITLE
WIP: Depend on executables, implements #213

### PR DESCRIPTION
**Pull request by iamed2**
Users can now declare `executable_dependency` and have it work just like `library_dependency`. 

Unfortunately as far as I can tell there is no easy way in Julia to check if a file is executable on Windows. However, I ensured that files with `PATHEXT` executable extensions are preferred. 

Currently `install` will write a variable `_exec` to deps.jl which contains a dictionary of dependency names and paths to executables. Please let me know if there is a better way. I mostly left `load_dependencies` alone, so it still gives tuples where the second element is the path. It's not clear whether including `deps.jl` or calling `load_dependencies` is preferred, and the results for doing either are undocumented. If someone knows the recommended way to approach this, I can document it. 

It's unclear to me how to test this new feature. Would the best approach be to just create and run a build script with an arbitrary executable dependency? The current approach seems to use build scripts from other packages, but since it's a new feature, no packages are using it. I have been testing with my build script for the unreleased EphemeralPG.

Work left:
- [x] Document `executable_dependency` ~~and the recommended way of loading dependencies~~
- [x] Adapt the Autotools build process for executable dependencies
- [x] Test
- [x] ~~Fix or~~ get around https://github.com/JuliaLang/Compat.jl/issues/187 (currently causing 0.4 failure)

This is here in WIP state so I can get comments sooner rather than later.

### CODE REVIEW

**Comment by tkelman** [comment]
The `load_dependencies` method is much less widely used than including `deps.jl`. The latter has the benefit that BinDeps is not required by the package at runtime, though precompilation mitigates the cost of that somewhat.

File: src/dependencies.jl [hunk 1]
```diff
@@ -15,20 +17,31 @@ type PackageContext
...
 end
 
-type LibraryGroup
+type DependencyGroup
     name::AbstractString
-    deps::Vector{LibraryDependency}
+    deps::Vector{Dependency}
 end
 
+typealias LibraryGroup DependencyGroup
```
Comment (nalimilan) [inline src/dependencies.jl]: I guess this is temporary (because of WIP) ?

File: src/dependencies.jl [hunk 2]
```diff
@@ -42,9 +55,31 @@ srcdir(dep) = joinpath(depsdir(dep),"src")
...
-successful_validate(l,p) = true
+successful_validate(l, p=nothing) = true
+
+function BinDeps.isexecutable(filepath)
+    if isfile(filepath)
+        res = false
+        @unix_only begin
+            res = ccall(:access, Cint, (Ptr{Uint8}, Cint), filepath, 1) == 0
+        end
+        # unfortunately there is no easy way to tell if this process can execute a file on Windows
```
Comment (nalimilan) [inline src/dependencies.jl]: Would there be a way to at least check whether the file is in an executable format?

File: src/dependencies.jl [hunk 3]
```diff
@@ -871,10 +1049,22 @@ macro install(_libmaps...)
...
+                        exec_deps = Dict()
                         for libkey in keys($libmaps)
-                            ((cached = get(load_cache,string(libkey),nothing)) === nothing) && continue
-                            println(depsfile, "@checked_lib ", $libmaps[libkey], " \"", escape_string(cached), "\"")
+                            cached = get(load_cache, string(libkey), nothing)
+                            if cached !== nothing
+                                (dep, path) = cached
+                                if isa(dep, BinDeps.LibraryDependency)
+                                    println(depsfile, "@checked_lib ", $libmaps[libkey], " \"", escape_string(path), "\"")
+                                else
```
Comment (nalimilan) [inline src/dependencies.jl]: Maybe do `elseif isa(dep, BinDeps.ExecutableDependency)` for clarity, and add a final `else` branch raising an error.

**Comment by nalimilan** [comment]
We should really recommend only one method and deprecate the others.

This looks good in general -- at least for what I understand since I don't really know that code.

> Currently install will write a variable _exec to deps.jl which contains a dictionary of dependency names and paths to executables. Please let me know if there is a better way.

What's the reason for the difference with libraries, for which we create separate variables?

File: src/dependencies.jl (hunk 2 above)
Comment (iamed2) [inline src/dependencies.jl]: The best we could do is ensure the file type ends with one of the PATHEXT extensions, and optionally check files that end in ".exe" for the executable marker (`MZ`). I'll implement that.

File: src/dependencies.jl (hunk 1 above)
Comment (iamed2) [inline src/dependencies.jl]: Yep

**Comment by iamed2** [comment]
Just realized the call overloading syntax I used has a 0.4 Compat fix but not a 0.3 one. Should I continue to support 0.3?

**Comment by nalimilan** [comment]
I'd say it's fine to drop 0.3 now. Anyway the package hasn't required many fixes in the recent months.

**Comment by iamed2** [comment]
Ok, I'll submit a separate PR for that.

**Comment by tkelman** [comment]
I think this has to be one of the last packages to drop 0.3 support. What's the Compat issue?

**Comment by iamed2** [comment]
Parametric call overloading for types (bug with kw args fixed in https://github.com/JuliaLang/Compat.jl/pull/188), which in v0.5 is just `(::Type{T}){T<:Supertype}(args...) = ...`. As I recall you just couldn't do that in 0.3 and there is no way to fake it such that behaviour is consistent with what is expected.

**Comment by tkelman** [comment]
Can you do `function dependency_constructor(T, context::PackageContext, name; properties...)` ?

**Comment by iamed2** [comment]
That's what I would have done had people wanted to continue 0.3 support.

**Comment by tkelman** [comment]
That didn't come up until 7 hours ago? If it were any package other than BinDeps or Compat I'd be much more in favor of dropping 0.3 support. But here the likelihood of having to backport bugfixes to fix some future issue and keep 0.3 working is somewhat high. Would making this feature support 0.3 be difficult enough to warrant the ongoing maintenance of needing a backport branch here?

**Comment by iamed2** [comment]
Nope, I'll just change it. I guess I didn't communicate very well the first time; I did not mean to imply that it was very difficult to support v0.3, I just wanted to know whether people cared about v0.3 still. Sounds like they do.

**Comment by tkelman** [comment]
A few packages are still actively tagging new versions that support 0.3, e.g. https://github.com/JuliaLang/METADATA.jl/pull/5088 just today. So I think the roots of the dependency tree need to wait until that stops being the case. Or we could press the issue with sufficient warning to the community, if it becomes worth maintaining multiple branches.

**Comment by iamed2** [comment]
New commit supports 0.3 and puts paths into variables rather than a dict. Added a test as well. I did not address the recommended way of loading dependencies, though I think `deps.jl` works better and leaks fewer abstractions.

**Comment by iamed2** [comment]
I've been trying to fix something on Windows for a while and I came across what seems like a bug. Seems like `install` only attempts to `satisfy!` with the platform's defaults, which means `SimpleBuild` will never run on Windows even if it satisfies the dependency.

**Comment by tkelman** [comment]
SimpleBuild has to be added to the defaults to work on Windows. Trying to compile anything on Windows users' systems at install time very rarely works.

**Comment by ararslan** [comment]
Bump. I'm currently in need of this functionality so I'd love to see this go live!

File: src/dependencies.jl [hunk 4]
```diff
@@ -42,9 +53,48 @@ srcdir(dep) = joinpath(depsdir(dep),"src")
...
 
-function _library_dependency(context::PackageContext, name; properties...)
+# normal contents of %PATHEXT% on Windows
+const win_pathext = (".COM", ".EXE", ".BAT", ".CMD", ".VBS", ".VBE", ".JS", ".JSE", ".WSF", ".WSH", ".MSC")
+const win_case_pathext = [collect(win_pathext); map!(lowercase, collect(win_pathext))]
+
+function BinDeps.isexecutable(filepath)
+    if isfile(filepath)
+        res = false
+        @unix_only begin
```
Comment (ararslan) [inline src/dependencies.jl]: Should use Compat and `@static is_*` these calls

**Comment by iamed2** [comment]
I couldn't get things working on all platforms. The testing infrastructure is just not there, and there is functionality partially implemented across many parts of the code, which makes it hard to change. At this point, I am abandoning this PR and waiting for BinDeps to be replaced (I was working on this for Invenia and given the difficulties I've encountered both using and editing BinDeps we've decided not to use it). Feel free to branch off this PR and continue the work.

File: src/dependencies.jl [hunk 5]
```diff
@@ -655,10 +870,10 @@ function can_provide(p::PackageManager,opts,dep)
     end
 end
 
-issatisfied(dep::LibraryDependency) = !isempty(_find_library(dep))
+issatisfied(dep::Dependency) = !isempty(_find_dependency(dep))
 
-allf(deps) = [dep => _find_library(dep) for dep in deps.deps]
-function satisfied_providers(deps::LibraryGroup, allfl = allf(deps))
+allf(deps) = [dep => _find_dependency(dep) for dep in deps.deps]
```
Comment (ararslan) [inline src/dependencies.jl]: This is deprecated syntax but IIRC there's no `@compat` for it. I think you can do
```julia
Dict([Pair(dep, _find_dependency(dep)) for dep in deps.deps])
```
to get around that?

**Comment by ararslan** [comment]
Ah, sad day. Regardless, thanks for your hard work on this! It's certainly inspiring.

Do you have any pointers as to how to declare executable dependencies if not through BinDeps? How is Inventia getting by without it?

**Comment by iamed2** [comment]
Thanks!

We're generally handling it outside of Julia by making containers/instances with what we require already installed, or having separate install scripts. In the case of https://github.com/invenia/FTPClient.jl we just bundled the dependencies in the repo. So we go in both directions from BinDeps; either require it to be installed, or install a specific version ourselves in the build script.

**Pull request closed without merging**